bellman/src/main.rs

```diff
@@ -0,0 +1,315 @@
+    let file = File::open(&filename).unwrap();
+    let mut result = Vec::new();
+    for line in BufReader::new(file).lines() {
+        for var in line.expect("Couldn't parse line").split_whitespace() {
+            result.push(F::from_str(var).unwrap());
+        }
+    }
+    result
+}
+
+fn add_parsed_lines<E: Engine>(mut lhs: ParsedLine<E>,rhs: &ParsedLine<E>) -> ParsedLine<E> {
+    assert!(lhs.index == rhs.index);
+    lhs.result.add_assign(&rhs.result);
+    ParsedLine {
+        index: lhs.index,
+        lc: lhs.lc + &rhs.lc,
+        debug: lhs.debug + " + " + &rhs.debug,
+        result: lhs.result,
+    }
+}
+
+struct CircuitImpl<F: PrimeField> {
+    primary_witness: Vec<F>,
+    aux_witness: Vec<F>,
+    qap_a: String,
+    qap_b: String,
+    qap_c: String,
+}
+
+impl<'a, E: Engine> Circuit<E> for &'a CircuitImpl<E::Fr> {
+    fn synthesize<CS: ConstraintSystem<E>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+        let now = Instant::now();
+
+        println!("  {} Reading Witness", now.elapsed().as_secs());
+        let aux: Vec<Assignment<E>> = self.aux_witness.clone().into_iter()
+            .map(|value| Assignment {
+                variable: cs.alloc(|| "aux", || Ok(value)).unwrap(), 
+                constrained: false, 
+                value: value
+            }).collect();
+        let primary: Vec<Assignment<E>> = self.primary_witness.clone().into_iter()
+            .map(|value| Assignment {
+                variable: cs.alloc_input(|| "aux", || Ok(value)).unwrap(), 
+                constrained: false, 
+                value: value
+            }).collect();
+
+        let mut vars = vec![
+            Assignment::<E>{
+                variable: CS::one(), 
+                constrained: false, 
+                value: E::Fr::one()
+            }];
+        vars.extend(aux);
+        vars.extend(primary);
+
+        println!("  {} Reading QAP", now.elapsed().as_secs());
+        {
+            let mut qap_a_file = BufReader::new(File::open(&self.qap_a)?);
+            let mut qap_b_file = BufReader::new(File::open(&self.qap_b)?);
+            let mut qap_c_file = BufReader::new(File::open(&self.qap_c)?);
+
+            let mut current_a_lc = empty_line(0);
+            let mut next_a_lc = read_next_constraint::<E>(&mut qap_a_file, &mut vars, false);
+
+            let mut current_b_lc = empty_line(0);
+            let mut next_b_lc = read_next_constraint::<E>(&mut qap_b_file, &mut vars, false);
+
+            let mut current_c_lc = empty_line(0);
+            let mut next_c_lc = read_next_constraint::<E>(&mut qap_c_file, &mut vars, true);
+
+            while next_a_lc.is_some() || next_b_lc.is_some() || next_c_lc.is_some() {
+                while next_a_lc.is_some() {
+                    let mut advance = false;
+                    {
+                        let next = next_a_lc.as_ref().unwrap();
+                        assert!(next.index >= current_a_lc.index);
```

Comment:
We will need to remove the `0 0 0` at the end of matrices `a` and `b`.

Reply:
Good catch. I'll just warn instead of asserting

Reply:
Shall I do the same in mine?

Reply:
I think that's more convenient than manually cleaning the pepper output.